cluster/kubernetes/kubernetes.go

```diff
@@ -483,6 +484,120 @@ func (c *Cluster) PublicSSHKey(regenerate bool) (ssh.PublicKey, error) {
 	return publicKey, nil
 }
 
+type servicePod struct {
+	s  v1.Service
+	pc podController
+}
+
+// Internal function to get a collection of service-controller pairs for a given namespace
+func (c *Cluster) allServices(ns string) (serviceControllers []servicePod, _ error) {
+	var namespaces []v1.Namespace
+	if ns == "" {
+		nsList, err := c.client.Namespaces().List(api.ListOptions{})
+		if err != nil {
+			return serviceControllers, errors.Wrap(err, "getting namespaces")
+		}
+		namespaces = nsList.Items
+	} else {
+		nsSingle, err := c.client.Namespaces().Get(ns)
+		if err != nil {
+			return serviceControllers, errors.Wrap(err, "getting namespaces")
+		}
+		namespaces = []v1.Namespace{*nsSingle}
+	}
+
+	// Foreach namespace
+	for _, ns := range namespaces {
+		services, err := c.client.Services(ns.Name).List(api.ListOptions{})
+		if err != nil {
+			return serviceControllers, errors.Wrapf(err, "getting services for namespace %s", ns.Name)
+		}
+
+		controllers, err := c.podControllersInNamespace(ns.Name)
+		if err != nil {
+			return serviceControllers, errors.Wrapf(err, "getting controllers for namespace %s", ns.Name)
+		}
+
+		// Foreach service
+		for _, service := range services.Items {
+			if isAddon(&service) {
+				continue
+			}
+
+			// Find controller for service
+			pc, err := matchController(&service, controllers)
+			if err != nil {
+				c.logger.Log(errors.Wrapf(cluster.ErrNoMatching, "matching controllers to service %s/%s", ns.Name, service.Name))
+				continue
+			}
+			serviceControllers = append(serviceControllers, servicePod{s: service, pc: pc})
+		}
+	}
+	return
+}
+
+// ImagesToFetch is a k8s specific method to get a list of images to update along with their credentials
+func (c *Cluster) ImagesToFetch() func() []registry.ImageCreds {
```

Comment:
You can safely make this return `[]registry.ImageCreds` and supply it as `k8s.ImagesToFetch` elsewhere, rather than returning the `func()`.